// change-case-commonjs/dist/keys.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.snakeCase = exports.sentenceCase = exports.pathCase = exports.pascalCase = exports.kebabCase = exports.noCase = exports.trainCase = exports.dotCase = exports.constantCase = exports.capitalCase = exports.camelCase = void 0;
const changeCase = __importStar(require("./index.js"));
const isObject = (object) => object !== null && typeof object === "object";
function changeKeysFactory(changeCase) {
    return function changeKeys(object, depth = 1, options) {
        if (depth === 0 || !isObject(object))
            return object;
        if (Array.isArray(object)) {
            return object.map((item) => changeKeys(item, depth - 1));
        }
        const result = Object.create(Object.getPrototypeOf(object));
        Object.keys(object).forEach((key) => {
            const value = object[key];
            const changedKey = changeCase(key, options);
            const changedValue = changeKeys(value, depth - 1);
            result[changedKey] = changedValue;
        });
        return result;
    };
}
exports.camelCase = changeKeysFactory(changeCase.camelCase);
exports.capitalCase = changeKeysFactory(changeCase.capitalCase);
exports.constantCase = changeKeysFactory(changeCase.constantCase);
exports.dotCase = changeKeysFactory(changeCase.dotCase);
exports.trainCase = changeKeysFactory(changeCase.trainCase);
exports.noCase = changeKeysFactory(changeCase.noCase);
exports.kebabCase = changeKeysFactory(changeCase.kebabCase);
exports.pascalCase = changeKeysFactory(changeCase.pascalCase);
exports.pathCase = changeKeysFactory(changeCase.pathCase);
exports.sentenceCase = changeKeysFactory(changeCase.sentenceCase);
exports.snakeCase = changeKeysFactory(changeCase.snakeCase);
